File: document/src/main/java/dev/sayaya/handbook/domain/Document.java
package dev.sayaya.handbook.domain;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;
import jsinterop.annotations.*;
import jsinterop.base.JsPropertyMap;
import lombok.*;
import lombok.experimental.Accessors;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
@JsType(isNative = true, namespace = JsPackage.GLOBAL, name = "Object")
@Getter(onMethod_ = {@JsOverlay, @JsIgnore})
@Setter(onMethod_ = {@JsOverlay, @JsIgnore})
@Accessors(fluent = true)
@NoArgsConstructor
public final class Document {
    @JsonProperty("id") @JsProperty private String id;
    @JsonProperty("type") @JsProperty private String type;
    @JsonProperty("serial") @JsProperty private String serial;
    @JsonProperty("effectDateTime") @JsProperty private double effectDateTime;
    @JsonProperty("expireDateTime") @JsProperty private double expireDateTime;
    @JsonProperty("createDateTime") @JsProperty private double createDateTime;
    @JsonProperty("creator") @JsProperty private String creator;
    @JsonProperty("data") @JsProperty private JsPropertyMap<String> data;
    @JsonProperty("status") @JsProperty private String status;
    @JsonProperty("rev") @JsProperty private long rev;

    @JsOverlay @JsIgnore
    public static Document create(String id, String type) {
        return create(id, type, null, 0.0, 0.0, 0.0, null, null);
    }

    @JsOverlay @JsIgnore
    public static Document create(String id, String type, String serial) {
        return create(id, type, serial, 0.0, 0.0, 0.0, null, null);
    }

    @JsOverlay @JsIgnore
    public boolean isExpired(double now) {
        return expireDateTime() > 0 && expireDateTime() <= now;
    }

    @JsOverlay @JsIgnore
    public static Document create(String id, String type, String serial, double effectDateTime, double expireDateTime, double createDateTime, String creator, jsinterop.base.JsPropertyMap<String> data) {
        if (serial != null && !serial.matches("^[a-zA-Z0-9_-]+$")) {
            throw new IllegalArgumentException("Document serial must be alphanumeric and may include hyphens and underscores.");
        }
        if (effectDateTime > 0 && expireDateTime > 0 && expireDateTime <= effectDateTime) {
            throw new IllegalArgumentException("Expire date time must be after effect date time");
        }
        if (id != null && (createDateTime <= 0 || creator == null)) {
            throw new IllegalArgumentException("If id is not null, createDateTime and creator must be not null");
        }
        Document doc = new Document();
        doc.id(id);
        doc.type(type);
        doc.serial(serial);
        doc.effectDateTime(effectDateTime);
        doc.expireDateTime(expireDateTime);
        doc.createDateTime(createDateTime);
        doc.creator(creator);
        doc.data(data);
        doc.status("DRAFT");
        doc.rev(-1);
        return doc;
    }

    @JsOverlay @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Document document = (Document) o;
        if (id() == null || document.id() == null) return false;
        return id().equals(document.id());
    }

    @JsOverlay @Override
    public int hashCode() {
        if (id() != null) return id().hashCode();
        return System.identityHashCode(this);
    }
}

File: document/src/main/kotlin/dev/sayaya/handbook/domain/Document.kt
package dev.sayaya.handbook.domain

import java.io.Serializable
import java.time.Instant
import java.util.*

/**
 * 시스템에서 사용되는 일반적인 문서를 나타내는 엔티티(Entity).
 *
 * 이 엔티티는 아직 영속화되지 않은 상태(id가 null)를 표현할 수 있습니다.
 * 엔티티의 동등성은 영속화된 후 할당되는 고유 식별자 `id`를 기준으로 판단합니다.
 */
@JvmRecord
data class Document (
    val id: UUID?,
    val type: String,
    val serial: String,
    val effectDateTime: Instant,
    val expireDateTime: Instant,
    val createDateTime: Instant?,
    val creator: String?,
    val data: Map<String, String?>,
): Serializable {
    init {
        require(serial.matches(Regex("^[a-zA-Z0-9-_]+$"))) { "Document serial must be alphanumeric and may include hyphens and underscores." }
        require(expireDateTime.isAfter(effectDateTime)) { "Expire date time must be after effect date time" }
        // ID가 있다는 것은 영속화되었음을 의미하므로, 생성 관련 메타데이터가 반드시 존재해야 함을 보장합니다.
        require(id == null || (createDateTime != null && creator != null)) { "If id is not null, createDateTime and creator must be not null" }
    }
    /**
     * 엔티티의 동등성 규칙을 적용하기 위해 재정의합니다.
     * - ID가 있는 객체: ID를 기준으로 비교합니다.
     * - ID가 없는 객체(영속화되지 않은 객체): 자기 자신 외에는 그 어떤 객체와도 같을 수 없습니다.
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Document
        // ID가 없는(null) 엔티티는 다른 어떤 객체와도 동일하다고 판단할 수 없습니다.
        if (id == null) return false
        return id == other.id
    }

    /**
     * equals가 id를 기준으로 동작하므로, hashCode도 id를 기반으로 생성합니다.
     * ID가 없는 객체는 기본 객체의 신원 해시코드(identity hash code)를 사용합니다.
     */
    override fun hashCode(): Int {
        return id?.hashCode() ?: System.identityHashCode(this)
    }
}